Regression `no-member` with inherited AnnAssign nodes

Regression from #7509
The false-positive is actually caused by an issue in astroid. In particular with https://github.com/PyCQA/astroid/blob/v2.12.11/astroid/nodes/scoped_nodes/scoped_nodes.py#L2570-L2577

Opened this so I don't forget to add the example as regression test once the issue is fixed.

```py
from __future__ import annotations

class Base:
    attr: int | None = None

class Parent(Base):
    attr: int

class Child(Parent):
    attr = 2

    def __init__(self):
        self.attr = self.attr | 4  # false-positive
```
```
test.py:13:20: E1101: Instance of 'Child' has no 'attr' member (no-member)
```

--- a/.github/workflows/tests.yaml
+++ b/.github/workflows/tests.yaml
@@ -191,7 +191,7 @@ jobs:
         run: >-
           echo "key=venv-${{ env.CACHE_VERSION }}-${{
             hashFiles('setup.cfg', 'requirements_test_min.txt')
-          }}" >> $GITHUB_OUTPUT
+          }}" >> $env:GITHUB_OUTPUT
       - name: Restore Python virtual environment
         id: cache-venv
         uses: actions/cache@v3.0.11

--- a/tests/functional/r/regression_02/regression_no_member_7631.py
+++ b/tests/functional/r/regression_02/regression_no_member_7631.py
@@ -1,0 +1,16 @@
+"""Regression test from https://github.com/PyCQA/pylint/issues/7631
+The following code should NOT raise no-member.
+"""
+# pylint: disable=missing-docstring,too-few-public-methods
+
+class Base:
+    attr: int = 2
+
+class Parent(Base):
+    attr: int
+
+class Child(Parent):
+    attr = 2
+
+    def __init__(self):
+        self.attr = self.attr | 4

